File: src/parser.cpp
#include "parser.h"
#include "visitor.h"

#include <iostream>

std::optional<Token> Parser::getCurrentToken() {
  if (token_idx_ >= tokens_.size()) {
    return std::nullopt;
  }
  return tokens_[token_idx_];
}

std::optional<Token> Parser::getNextToken() {
  advance();
  return getCurrentToken();
}

Token Parser::expectedNextToken(TokenType type) {
  auto token = getNextToken();
  if (!token) {
    std::cout << "Expecting next token " << type << " but didn't find one"
              << std::endl;
    exit(1);
  }

  if (token->getType() != type) {
    std::cout << "Expected token type " << type << " at " << token_idx_
              << ", but got " << token->getType() << std::endl;
    exit(1);
  }

  return *token;
}

void Parser::advance() { token_idx_++; }

std::unique_ptr<ExprNode> Parser::handlePrimary() {
  auto token = getCurrentToken();
  if (!token) {
    std::cout << "Expected token in expression";
    exit(1);
  }

  switch (token->getType()) {
  case tok_identifier: {
    // peek ahead to see if its a fn call or a variable
    bool is_fn_call = token_idx_ + 1 < tokens_.size() &&
                      tokens_[token_idx_ + 1].getType() == tok_lpar;
    if (is_fn_call) {
      std::vector<std::unique_ptr<ExprNode>> args;
      expectedNextToken(tok_lpar);
      while (auto token = getNextToken()) {
        if (token->getType() != tok_identifier) {
          break;
        }
        args.push_back(handleExpression());
        if (!getCurrentToken() || getCurrentToken()->getType() != tok_comma) {
          break;
        }
        expectedNextToken(tok_comma);
      }
      advance(); // skip past rpar
      return std::make_unique<FunctionCallExprNode>(token->getIdentifier(),
                                                    std::move(args));
    }
    advance();
    return std::make_unique<IdentifierExprNode>(token->getIdentifier());
  }
  case tok_number:
    advance();
    return std::make_unique<NumberLiteralNode>(token->getNumber());
  case tok_lpar: {
    advance();
    auto expr = handleExpression();
    if (!getCurrentToken() || getCurrentToken()->getType() != tok_rpar) {
      std::cout << "error when handling parenthesized expression, got token "
                << getCurrentToken()->getType()
                << " expected tok_rpar, at token idx " << token_idx_
                << std::endl;
      exit(1);
    }
    advance();
    return expr;
  }
  default:
    std::cout << "unexpected token when parsing primary expr" << std::endl;
    exit(1);
  }
}

int32_t Parser::getBinOpPrecedence(Token bin_op) {
  switch (bin_op.getType()) {
  case tok_add:
    return 0;
  case tok_sub:
    return 1000;
  case tok_mul:
    return 2000;
  case tok_div:
    return 3000;
  case tok_mod:
    return 4000;
  case tok_gt:
    return 5000;
  case tok_gte:
    return 6000;
  case tok_lt:
    return 7000;
  case tok_lte:
    return 8000;
  default:
    return -1;
  }
}
std::unique_ptr<ExprNode>
Parser::handleBinOpRHS(int8_t precedence, std::unique_ptr<ExprNode> LHS) {
  while (true) {

    auto bin_op = getCurrentToken();

    if (!bin_op) {
      std::cout << "Expected token in expression";
      exit(1);
    }

    int32_t bin_op_precedence = getBinOpPrecedence(*bin_op);
    if (getBinOpPrecedence(*bin_op) < precedence) {
      return LHS;
    }
    advance();

    auto RHS = handlePrimary();
    if (!RHS) {
      std::cout << "Expected expression after binary operator" << std::endl;
      exit(1);
    }

    auto next_bin_op = getCurrentToken();
    if (!next_bin_op) {
      return std::make_unique<BinaryExprNode>(bin_op->getType(), std::move(LHS),
                                              std::move(RHS));
    }

    int32_t next_bin_op_precedence = getBinOpPrecedence(*next_bin_op);
    if (bin_op_precedence < next_bin_op_precedence) {
      RHS = handleBinOpRHS(next_bin_op_precedence + 1, std::move(RHS));
    }

    LHS = std::make_unique<BinaryExprNode>(bin_op->getType(), std::move(LHS),
                                           std::move(RHS));
  }
}

std::unique_ptr<ExprNode> Parser::handleExpression() {
  auto primary = handlePrimary();
  return handleBinOpRHS(0, std::move(primary));
}

std::unique_ptr<ConditionalNode> Parser::handleConditional() {
  advance(); // skip if token
  auto if_expr = handleExpression();

  advance(); // skip lbrak
  auto if_body = handleBody();
  advance(); // skip rbrak
  if (auto token = getCurrentToken()) {
    if (token->getType() == tok_else) {
      expectedNextToken(tok_lbrak);
      advance(); // skip lbrak
      auto else_body = handleBody();
      advance(); // skip rbrak
      return std::make_unique<ConditionalNode>(
          std::move(if_expr), std::move(if_body), std::move(else_body));
    }
  }

  return std::make_unique<ConditionalNode>(std::move(if_expr),
                                           std::move(if_body));
}

std::unique_ptr<ReturnNode> Parser::handleReturnStatement() {
  //   expectedNextToken(tok_return);
  advance(); // skip return
  return std::make_unique<ReturnNode>(std::move(handleExpression()));
}

std::unique_ptr<DefinitionNode> Parser::handleDefinition() {
  auto lvalue = getCurrentToken();
  expectedNextToken(tok_equals);
  advance();
  auto expr = handleExpression();
  return std::make_unique<DefinitionNode>(lvalue->getIdentifier(),
                                          std::move(expr));
}

std::unique_ptr<BodyNode> Parser::handleBody() {
  std::vector<std::unique_ptr<BodySubNode>> blocks;
  while (auto token = getCurrentToken()) {
    if (!token) {
      std::cout << "Expected another token in function body" << std::endl;
      exit(1);
    }

    switch (token->getType()) {
    case tok_if: {
      blocks.push_back(handleConditional());
      break;
    }
    case tok_return: {
      blocks.push_back(handleReturnStatement());
      break;
    }
    case tok_identifier: {
      blocks.push_back(handleDefinition());
      break;
    }
    case tok_rbrak: {
      return std::make_unique<BodyNode>(std::move(blocks));
    }
    default: {
      std::cout << "Expected conditional, return statemnet, or identifier in "
                   "function body but got "
                << token->toString() << " at " << token_idx_ << std::endl;
      exit(1);
    }
    }
    if (getCurrentToken() && getCurrentToken()->getType() == tok_rbrak) {
      break;
    }
  }
  return std::make_unique<BodyNode>(std::move(blocks));
}

std::unique_ptr<FunctionNode> Parser::handleFunction() {
  // consume function name
  std::optional<Token> fnName = getNextToken();
  if (!fnName || (*fnName).getType() != tok_identifier) {
    std::cout << "Error, expecting identifer in function declaration"
              << std::endl;
  }

  // consume lparenthesis
  expectedNextToken(tok_lpar);

  auto token = getNextToken();
  std::vector<std::string> args;
  while (token && token->getType() != tok_rpar) {
    if (token->getType() == tok_identifier) {
      args.push_back(token->getIdentifier());
      advance();

      if (!getCurrentToken()) {
        std::cout << "Error when parsing function declaration, expected comma"
                  << std::endl;
        exit(1);
      }

      token.emplace(*getCurrentToken());
      if (token->getType() == tok_rpar) {
        break;
      }

      if (token->getType() != tok_comma) {
        std::cout << "Error when parsing function declaration, expected comma"
                  << std::endl;
      }
      advance();

    } else {
      std::cout << "Error when parsing function declaration" << std::endl;
      exit(1);
    }
  }

  auto functionDeclaration = std::make_unique<FunctionDeclarationNode>(
      fnName->getIdentifier(), std::move(args));

  expectedNextToken(tok_lbrak);
  advance(); // skip lbrak
  auto body = handleBody();
  advance(); // skip rbrak
  return std::make_unique<FunctionNode>(std::move(functionDeclaration),
                                        std::move(body));
}

std::unique_ptr<Program> Parser::parse() {
  std::vector<std::unique_ptr<FunctionNode>> functions;
  while (auto token = getCurrentToken()) {
    switch (token->getType()) {
    case tok_eof:
      return std::make_unique<Program>(std::move(functions));
    case tok_def:
      functions.push_back(std::move(handleFunction()));
      break;
    default:
      std::cout << "Error: received invalid token type " << token->toString()
                << std::endl;
      exit(1);
    }
  }
  return std::make_unique<Program>(std::move(functions));
}

void FunctionDeclarationNode::accept(Visitor *v) {
  v->visitFunctionDeclarationNode(this);
}
void BinaryExprNode::accept(Visitor *v) { v->visitBinaryExprNode(this); }
void NumberLiteralNode::accept(Visitor *v) { v->visitNumberLiteralNode(this); }
void IdentifierExprNode::accept(Visitor *v) {
  v->visitIdentifierExprNode(this);
}
void FunctionCallExprNode::accept(Visitor *v) {
  v->visitFunctionCallExprNode(this);
}
void BodyNode::accept(Visitor *v) { v->visitBodyNode(this); }
void ConditionalNode::accept(Visitor *v) { v->visitConditionalNode(this); }
void DefinitionNode::accept(Visitor *v) { v->visitDefinitionNode(this); }
void ReturnNode::accept(Visitor *v) { v->visitReturnNode(this); }
void FunctionNode::accept(Visitor *v) { v->visitFunctionNode(this); }
void Program::accept(Visitor *v) { v->visitProgramNode(this); }

File: src/parser.rs
use crate::scanner::{Token, TokenType};
use std::vec::Vec;

trait Visitable {
    fn visit(tokens: &mut TokenStream) -> Self;
}

pub struct TokenStream {
    tokens: Vec<Token>,
}

impl TokenStream {
    pub fn new(tokens: Vec<Token>) -> TokenStream {
        TokenStream { tokens }
    }

    fn peek_next_token(&self) -> Option<&Token> {
        if self.tokens.len() > 1 {
            return Some(&self.tokens[1]);
        }
        None
    }

    fn expected_current_token(&mut self, expected_token_type: TokenType) {
        if let Some(curr_token) = self.get_current_token() {
            // TODO: implement fmt::Display for TokenType
            assert!(
                curr_token.token_type == expected_token_type,
                "expected {}, actual {}",
                Token::new(expected_token_type, String::from(""), 0.0),
                curr_token
            );
        } else {
            panic!(
                "Expected token type {} but didn't find it",
                Token::new(expected_token_type, String::from(""), 0.0)
            );
        }
    }

    fn consume(&mut self) -> Option<Token> {
        println!("Parsed token {}", self.tokens[0]);
        // TODO: this is inefficient, swap it out with vecdeq
        if self.tokens.len() > 0 {
            Some(self.tokens.remove(0))
        } else {
            None
        }
    }

    fn remove_current_token(&mut self) -> Option<Token> {
        if self.tokens.len() > 0 {
            Some(self.tokens.remove(0))
        } else {
            None
        }
    }

    fn get_current_token(&self) -> Option<&Token> {
        if self.tokens.len() == 0 {
            None
        } else {
            Some(&self.tokens[0])
        }
    }
}

pub struct FunctionDeclaration {
    name: String,
    args: Vec<String>,
}

impl Visitable for FunctionDeclaration {
    fn visit(tokens: &mut TokenStream) -> FunctionDeclaration {
        tokens.expected_current_token(TokenType::TokDef); // skipping 'def' token
        tokens.consume();

        tokens.expected_current_token(TokenType::TokIdentifier);
        let fn_name = tokens.consume();
        if fn_name.is_none() {
            panic!("Invalid token after start of function definition");
        }

        tokens.expected_current_token(TokenType::TokLpar);
        tokens.consume();

        // get function args
        let mut args: Vec<String> = vec![];
        while let Some(token) = tokens.consume() {
            if token.token_type == TokenType::TokRpar {
                break;
            }
            if token.token_type == TokenType::TokIdentifier {
                args.push(token.identifier);
                if let Some(current_token) = tokens.get_current_token() {
                    if current_token.token_type == TokenType::TokRpar {
                        break;
                    }
                    tokens.expected_current_token(TokenType::TokComma);
                    tokens.consume();
                } else {
                    panic!("Expected a token in function declaration argument list");
                }
            } else {
                panic!("Failed to parse function declaration {}", token);
            }
        }

        tokens.expected_current_token(TokenType::TokRpar);
        tokens.consume();
        FunctionDeclaration {
            name: fn_name.unwrap().identifier,
            args,
        }
    }
}

pub enum Expression {
    Binary(BinaryExpression),
    NumberLiteral(NumberLiteralExpression),
    Identifier(IdentifierExpression),
    FunctionCall(FunctionCallExpression),
}

pub struct BinaryExpression {}
impl Visitable for BinaryExpression {
    fn visit(tokens: &mut TokenStream) -> Self {
        todo!("Implement binary expression");
    }
}

pub struct NumberLiteralExpression {}
impl Visitable for NumberLiteralExpression {
    fn visit(tokens: &mut TokenStream) -> Self {
        todo!("Implement number literal expression");
    }
}

pub struct IdentifierExpression {
    name: String,
}
impl Visitable for IdentifierExpression {
    fn visit(tokens: &mut TokenStream) -> Self {
        tokens.expected_current_token(TokenType::TokIdentifier);
        if let Some(identifier_token) = tokens.consume() {
            IdentifierExpression {
                name: identifier_token.identifier,
            }
        } else {
            panic!("Expected token in identifier expression but didnt find one");
        }
    }
}

pub struct FunctionCallExpression {}
impl Visitable for FunctionCallExpression {
    fn visit(tokens: &mut TokenStream) -> Self {
        todo!("Implement number literal expression");
    }
}

impl Expression {
    fn handleLHS(tokens: &mut TokenStream) -> Expression {
        if let Some(token) = tokens.get_current_token() {
            match token.token_type {
                TokenType::TokIdentifier => {
                    if let Some(identifier_token) = tokens.get_current_token() {
                        if let Some(next_token) = tokens.peek_next_token() {
                            // token after means it can be a variable or a function call
                            let is_fn_call = next_token.token_type == TokenType::TokLpar;
                            if is_fn_call {
                                return Expression::FunctionCall(FunctionCallExpression::visit(
                                    tokens,
                                ));
                            } else {
                                return Expression::Identifier(IdentifierExpression::visit(tokens));
                            }
                        } else {
                            // no token after the identifier means its just an identifier
                            // expression
                            return Expression::Identifier(IdentifierExpression::visit(tokens));
                        }
                    } else {
                        panic!("Expected identifier token but didn't find one");
                    }
                }
                _ => panic!("Received unexpected token type in expression"),
            }
        } else {
            panic!("Expected token in expression but didn't find one");
        }
    }

    fn handleBinOpRHS(tokens: &mut TokenStream, lhs: Expression) -> Expression {
        todo!("handleBinOpRHS");
    }
}

impl Visitable for Expression {
    fn visit(tokens: &mut TokenStream) -> Expression {
        let lhs = Expression::handleLHS(tokens);
        Expression::handleBinOpRHS(tokens, lhs)
    }
}

enum ScopeStatement {
    Definition(DefinitionScopeStatement),
    Conditional(ConditionalScopeStatement),
    Return(ReturnScopeStatement),
}
pub struct DefinitionScopeStatement {
    lvalue_identifier: String,
    rvalue_expression: Expression,
}
pub struct ReturnScopeStatement {}
pub struct ConditionalScopeStatement {}

impl Visitable for DefinitionScopeStatement {
    fn visit(tokens: &mut TokenStream) -> DefinitionScopeStatement {
        if let Some(lvalue_token) = tokens.consume() {
            tokens.expected_current_token(TokenType::TokEquals);
            tokens.consume();
            DefinitionScopeStatement {
                lvalue_identifier: lvalue_token.identifier,
                rvalue_expression: Expression::visit(tokens),
            }
        } else {
            panic!("Expected a token in definition but didn't find one");
        }
    }
}

impl Visitable for ReturnScopeStatement {
    fn visit(tokens: &mut TokenStream) -> ReturnScopeStatement {
        todo!("Need to implement return statement");
        ReturnScopeStatement {}
    }
}

impl Visitable for ConditionalScopeStatement {
    fn visit(tokens: &mut TokenStream) -> ConditionalScopeStatement {
        todo!("Need to implement conditional statement");
        tokens.expected_current_token(TokenType::TokIf);
        ConditionalScopeStatement {}
    }
}

pub struct Scope {}
impl Visitable for Scope {
    fn visit(tokens: &mut TokenStream) -> Scope {
        tokens.expected_current_token(TokenType::TokLbrak);
        tokens.consume();

        let mut blocks: Vec<ScopeStatement> = Vec::new();
        while let Some(token) = tokens.get_current_token() {
            match token.token_type {
                TokenType::TokIf => blocks.push(ScopeStatement::Conditional(
                    ConditionalScopeStatement::visit(tokens),
                )),
                TokenType::TokReturn => {
                    blocks.push(ScopeStatement::Return(ReturnScopeStatement::visit(tokens)))
                }
                TokenType::TokIdentifier => blocks.push(ScopeStatement::Definition(
                    DefinitionScopeStatement::visit(tokens),
                )),
                TokenType::TokRbrak => break,
                _ => panic!("Found unexpected token in scope"),
            }
        }
        tokens.expected_current_token(TokenType::TokRbrak);
        tokens.consume();
        Scope {}
    }
}

pub struct Function {
    declaration: FunctionDeclaration,
    body: Scope,
}

impl Visitable for Function {
    fn visit(tokens: &mut TokenStream) -> Function {
        let declaration = FunctionDeclaration::visit(tokens);
        let body = Scope::visit(tokens);
        Function { body, declaration }
    }
}

pub struct Program {
    functions: Vec<Function>,
}

impl Program {
    pub fn new(functions: Vec<Function>) -> Program {
        Program { functions }
    }
}

pub struct Parser {
    tokens: TokenStream,
}

impl Parser {
    pub fn new(tokens: Vec<Token>) -> Parser {
        Parser {
            tokens: TokenStream::new(tokens),
        }
    }

    pub fn parse(&mut self) -> Program {
        let mut functions = vec![];
        while let Some(token) = self.tokens.get_current_token() {
            match token.token_type {
                TokenType::TokEof => return Program::new(functions),
                TokenType::TokDef => {
                    let function = Function::visit(&mut self.tokens);
                    functions.push(function)
                }
                _ => panic!("Invalid token found"),
            }
        }
        return Program::new(functions);
    }
}
